playshogi-library-shogi/src/main/java/com/playshogi/library/shogi/rules/ShogiRulesEngine.java

```diff
@@ -241,30 +260,29 @@ private boolean isCaptureMoveLegalInPosition(final ShogiPosition position, final
 
 
         for (Square everySquare : position.getAllSquares()) { //for every square on the board
-            if (position.getPieceAt(everySquare) != null && position.getPieceAt(everySquare).isSentePiece() == isSente) {
+            Piece piece = position.getPieceAt(everySquare);
+            if (piece != null && piece.isSentePiece() == isSente) {
                 //check if there is sente's piece
                 List<Square> targetSquares = getPossibleTargetSquares(position, everySquare);
                 // find its possible squares to move
                 for (Square targetSquare : targetSquares) { //add each of those squares as a possible move
-                    if (position.getPieceAt(targetSquare) != null) { //check if it is a capturing move
-                        result.add(new CaptureMove(position.getPieceAt(everySquare), everySquare, targetSquare,
-                                false, position.getPieceAt(targetSquare)));
-                        if (canMoveWithPromotion(position, result.get(result.size() - 1))) {//if can promote,
-                            // capture
-                            // with promotion too
-                            result.add(new CaptureMove(position.getPieceAt(everySquare), everySquare,
-                                    targetSquare, true, position.getPieceAt(targetSquare)));
+                    Piece targetPiece = position.getPieceAt(targetSquare);
+                    if (targetPiece != null) { //check if it is a capturing move
+                        NormalMove capture = new CaptureMove(piece, everySquare, targetSquare, false, targetPiece);
+                        if (canMoveWithUnpromotion(position, capture)) {
```

Comment:
Shouldn't there be a branch "Piece is already promoted || canMoveWithUnpromotion" ? Since for a promoted piece, it is always possible to move without promotion.

Alternatively, we could replace "canMoveWithUnpromotion" with "canMoveWithoutPromoting", which would return true when the piece is promoted.

Reply:
I like "canMoveWithoutPromoting" and looking closer, found a way to make NormalMove and CaptureMove immutable and (at least for me) make the promoted/promotion concepts either to understand.